DOCSP-9560 crud skip (#67)
mongodb/docs-java: source/fundamentals/crud/read-operations/skip.txt

=====================
Skip Returned Results
=====================

.. default-domain:: mongodb

In this guide, we show you how to skip a specified number of returned results.

You can skip results on the returned results of a query by using the
``skip()`` method. You can also skip documents at a specific stage in an
aggregation pipeline by specifying a ``$skip`` aggregation stage.

The ``skip()`` method takes an integer that specifies the number of documents
to omit from the beginning of the list of documents returned by the
:java-docs:`FindIterable </apidocs/mongodb-driver-sync/com/mongodb/client/FindIterable.html#skip(int)>`. 

You can use the ``skip()`` method to skip the first two documents as follows:

.. code-block:: java

   collection.find().skip(2); 

:java-docs:`Aggregates.skip() </apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#skip(int)>`

> is an optional stage in the aggregation pipeline to specify how many

Reviewer: ```suggestion
is an optional stage in the aggregation pipeline that specifies how many
```

   { "_id": 6, "color": "pink", "qty": 3 }
   { "_id": 7, "color": "green", "qty": 8 }
   { "_id": 8, "color": "orange", "qty": 7 }

To address the scenario, the paint store needs to query the
``paint_inventory`` collection with an empty filter, sort the documents
by ``qty`` field and omit the first five results.

Using the ``skip()`` method:
````````````````````````````

.. literalinclude:: /includes/fundamentals/code-snippets/Skip.java
   :language: java
   :dedent:
   :start-after: begin skipExample
   :end-before: end skipExample

- The ``find()`` method returns all documents.  
- The ``sort()`` method specifies documents to display from highest to lowest based on the ``qty`` field.
- The ``skip()`` method specifies to omit the first 5 documents. 

Using ``Aggregates.skip()``:
````````````````````````````

.. literalinclude:: /includes/fundamentals/code-snippets/Skip.java
   :language: java
   :dedent:
   :start-after: begin skipAggregateExample
   :end-before: end skipAggregateExample

- The ``match()`` stage returns all documents.  
- The ``sort()`` stage specifies documents to display from highest to lowest based on the ``qty`` field.

> - The ``skip()`` stage specifies to omit the first 5 documents. 

Reviewer: Same comment about numbers https://github.com/mongodb/docs-java/pull/67#discussion_r603282823

The following shows the output of both queries specified above:

.. code-block:: json
   :copyable: false
   
   { "_id": 4, "color": "white", "qty": 6 }
   { "_id": 1, "color": "red", "qty": 5 }
   { "_id": 6, "color": "pink", "qty": 3 }

> After we run the query, we find the three best-selling colors are pink,

Reviewer: Same comment about switching perspective: https://github.com/mongodb/docs-java/pull/67#discussion_r603573400

red, and white. 

.. note::
   If the value of skip is greater than or equal to the number of matched
   documents for a query, that query returns no documents.

   If the ``skip()`` method from the example above skips the first nine
   documents, no results would return since there are only eight

>    documents. 

Reviewer: Wouldn't it be easier for the reader to point them directly to the reason you stated earlier?
"skips the first nine documents. Since the specified quantity exceeds the number of matched documents, the operation returns no results."

=====================
Skip Returned Results
=====================

.. default-domain:: mongodb

Use ``skip()`` to specify how many documents to omit from the beginning

> of the list of returned documents for a read operation. You can combine

Reviewer: Can you use skip on an aggregation? And do aggregations fit in the classification of the aforementioned read operations? Could you provide info about that here?

``skip()`` with ``sort()`` to omit the top (for descending order) or

> bottom (for ascending order) results for a given query. Since the order

Author: It is! https://docs.mongodb.com/drivers/node/current/fundamentals/crud/read-operations/skip/

> bottom (for ascending order) results for a given query. Since the order

Reviewer: Cool. We should rely on the server documentation for terminology when possible. I think "top" and "bottom" are uncommon terms and therefore you ended up explaining them in parentheses, but feel free to keep it that way if there isn't a better way to describe it.

   { "_id": 2, "color": "purple", "qty": 10 }
   { "_id": 3, "color": "blue", "qty": 9 }
   { "_id": 4, "color": "white", "qty": 6 }
   { "_id": 5, "color": "yellow", "qty": 11 }
   { "_id": 6, "color": "pink", "qty": 3 }
   { "_id": 7, "color": "green", "qty": 8 }
   { "_id": 8, "color": "orange", "qty": 7 }

To address the scenario, we query the ``paint`` collection with an empty
filter, sort the documents by ``qty`` field and omit the first 5
results.

.. literalinclude:: /includes/fundamentals/code-snippets/Skip.java
   :language: java
   :dedent:
   :start-after: begin skipExample
   :end-before: end skipExample

The find() method returns all documents.  The sort() method specifies
documents should appear highest to lowest based on the ``qty`` field.
The skip() method specifies that the first 5 documents should not appear
from the query. 

The following shows the output of the query specified above:

.. code-block:: json
   :copyable: false
   
   { "_id": 4, "color": "white", "qty": 6 }
   { "_id": 1, "color": "red", "qty": 5 }
   { "_id": 6, "color": "pink", "qty": 3 }

To answer the customer, the 3 best selling colors are: pink, red, and
white.

.. note::

>     If the value of skip exceeds the number of matched documents for a

Author: You're right, it was off by one, but no warnings/errors were in the build log.

=====================
Skip Returned Results
=====================

.. default-domain:: mongodb

> In this guide, we show you how to skip returned results.

Reviewer: I think it could be helpful to be more specific here.

```suggestion
In this guide, we show you how to skip a specified number of returned results.
```

You can skip results using the ``skip()`` method of a 

> :java-docs:`FindIterable </apidocs/mongodb-driver-sync/com/mongodb/client/FindIterable.html#skip(int)>`

Reviewer: Some users might not know what a `FindIterable` is, so we should provide a higher-level explanation. 
E.g. "You can specify the number of results to skip by calling the skip() method on the result of a query or in an aggregate pipeline."
Then, you can introduce the FindIterable as the result of the find() query on line 18 and something similar on line 24.

, so it could be better to specify the exact types and link to API documentation in the later.

instance or by using the :java-docs:`Aggregates.skip() </apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#sort(org.bson.conversions.Bson)>` 
static factory method in an aggregation pipeline. 

The ``skip()`` method takes an integer that specifies how many documents
to omit from the beginning of the list of documents returned by your
query. 

> You can use the ``skip()`` method of a ``FindIterable`` instance as follows:

Reviewer: Could you explain the "2" value in the example? I think it would help the reader understand its usage without having to dive into the API documentation.

.. code-block:: java

   collection.find().skip(2); 

The ``Aggregates.skip()`` method implements the ``Bson`` interface to

> specify skip criteria as an argument.  

Reviewer: 1) Since methods don't/can't implement interfaces, we shouldn't make that claim.

2) Usually, when you use a sentence in the structure "___ does Y to do Z", the Y is the precedent for the consequence Z. Is "[implementing the Bson interface]" the direct cause of "[specifying the skip criteria as an argument]"? Maybe you intended to list two different items in which you want to use a conjunction like "and" or "or" instead of a preposition like "to". Let me know if this makes sense.

> You can use the ``Aggregates.sort()`` method as follows:

Reviewer: I think you meant skip here.

.. code-block:: java

   import com.mongodb.client.model.Aggregates;

   collection.aggregate(Arrays.asList(Aggregates.skip(2)));
   

> The example in this guide is about a paint store that sells eight

Reviewer: Nitpick: It could be helpful here to add a header here to identify that you're starting a new section of material.

different colors of paint.  The best colors sell quicker than the other

> colors. One day, a customer comes in and asks what the three

Reviewer: Nitpick: omit unnecessary details.
Also one of the fixes included in: https://github.com/mongodb/docs-java/pull/67#discussion_r601986915

```suggestion
colors. One day, a customer asks what the three
```

best-selling colors are.  The following documents are in the

> ``paint_inventory`` collection: 

Reviewer: This abrupt transition and lack of explanation of "qty" is still bothering me (mentioned in https://github.com/mongodb/docs-java/pull/67#pullrequestreview-621736838 )

.. code-block:: json

   { "_id": 1, "color": "red", "qty": 5 }
   { "_id": 2, "color": "purple", "qty": 10 }
   { "_id": 3, "color": "blue", "qty": 9 }
   { "_id": 4, "color": "white", "qty": 6 }
   { "_id": 5, "color": "yellow", "qty": 11 }
   { "_id": 6, "color": "pink", "qty": 3 }
   { "_id": 7, "color": "green", "qty": 8 }
   { "_id": 8, "color": "orange", "qty": 7 }

To address the scenario, you query the ``paint_inventory`` collection

> with an empty filter, sort the documents by ``qty`` field and omit the first 5

Reviewer: Could you update all instances of numbers under 10 that aren't spelled out unless they're values or required to be in that format?

https://github.com/mongodb/docs-java/pull/67#discussion_r601987039

```suggestion
with an empty filter, sort the documents by ``qty`` field and omit the first five
```

results.

.. literalinclude:: /includes/fundamentals/code-snippets/Skip.java
   :language: java
   :dedent:
   :start-after: begin skipExample
   :end-before: end skipExample

- The ``find()`` method returns all documents.  
- The ``sort()`` method specifies documents to display from highest to lowest based on the ``qty`` field.
- The ``skip()`` method specifies to omit the first 5 documents. 

The following shows the output of the query specified above:

.. code-block:: json
   :copyable: false
   
   { "_id": 4, "color": "white", "qty": 6 }
   { "_id": 1, "color": "red", "qty": 5 }
   { "_id": 6, "color": "pink", "qty": 3 }

> After we run the query, we found the three best-selling colors are pink,

Reviewer: Maintain the same tense.
```suggestion
After we run the query, we find the three best-selling colors are pink,
```

red, and white. 

.. note::
   If the value of skip exceeds the number of matched documents for a

>    query, that query returns no documents.

Reviewer: Nitpick: it wouldn't hurt to also show the syntax of the equivalent aggregation to retrieve these results.

.. default-domain:: mongodb

In this guide, we show you how to skip a specified number of returned results.

You can skip results using the ``skip()`` method on a result of a query or
in an aggregation pipeline. 

The ``skip()`` method takes an integer that specifies the number of documents
to omit from the beginning of the list of documents returned by the
:java-docs:`FindIterable </apidocs/mongodb-driver-sync/com/mongodb/client/FindIterable.html#skip(int)>`. 

You can use the ``skip()`` method to skip the first two documents as follows:

.. code-block:: java

   collection.find().skip(2); 

The :java-docs:`Aggregates.skip() </apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html#skip(int)>`

> is a stage in the aggregation pipeline where you can specify a skip criteria as an argument.  

Reviewer: Aggregates.skip() is a builder method, not a stage. You can use Aggregates.skip() to build the Bson that represents an aggregation stage (you can also use other objects like Document or JsonObject), so we shouldn't refer to it as "a stage in the aggregation pipeline". Maybe something like "you can include Aggregates.skip() as a stage in the aggregation pipeline to specify how many documents to omit from the beginning of the results of the prior stage."
